### PR TITLE
Magic Link Authenticator

### PR DESCRIPTION
**Pull request by thukaraka**
Magic Link  local authenticator for wso2 identity server

### CODE REVIEW

File: src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+    private static final long serialVersionUID = 4345354156955223654L;
+    private static final Log log = LogFactory.getLog(MagicLinkAuthenticator.class);
+
+    /**
+     * @param request  The httpServletRequest.
+     * @param response The httpServletResponse.
+     * @param context  The authentication context.
+     * @throws AuthenticationFailedException In occasions of failing to send the magicToken to the user.
+     */
```
Comment (sumedhe) [inline src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java:57]: Let's add a description to the doc comment.

File: src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/JWTValidator.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+     * @return true for a JWT token.
+     */
+    private static boolean isJWT(String tokenIdentifier) {
+
+        return StringUtils.countMatches(tokenIdentifier, DOT_SEPARATOR) == 2;
+    }
+}
+
+
+
```
Comment (sumedhe) [inline src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/JWTValidator.java:156]: Unnecessary new lines. Shall we remove these?

File: src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+        return MagicLinkAuthenticatorConstants.AUTHENTICATOR_FRIENDLY_NAME;
+    }
+
+    @Override
+    public boolean canHandle(HttpServletRequest httpServletRequest) {
+
+        return StringUtils.isNotEmpty(httpServletRequest.getParameter("mL")) &&
+                StringUtils.isEmpty(httpServletRequest.getParameter("mL"))
+                || StringUtils.isNotEmpty(httpServletRequest.getParameter("mL"));
```
Comment (sumedhe) [inline src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java:151]: Let's create a constant for parameter key.

File: README.md [hunk 4]
```diff
@@ -1,2 +1,262 @@
...
+
+
+
+
+
+
+ 
+ 
+ 
+
```
Comment (sumedhe) [inline README.md:262]: Let's remove these empty lines.

**Review by sumedhe — APPROVED** [review]

**Review by janakamarasena — APPROVED** [review]

**Pull request merged**